src/engine/Sort.cpp

```diff
@@ -54,121 +58,148 @@ std::string Sort::getDescriptor() const {
 
 // _____________________________________________________________________________
 Result Sort::computeResult(bool requestLaziness) {
-  // Check if we should use external sorting.
-  bool useExternalSort =
-      getRuntimeParameter<&RuntimeParameters::sortExternal_>();
-
-  // Get the input lazily if we want to do external sorting, and materialized
-  // otherwise.
-  std::shared_ptr<const Result> subRes = subtree_->getResult(useExternalSort);
-
-  // Do the sorting, either external or in-memory.
-  if (useExternalSort) {
-    return computeResultExternal(subRes, requestLaziness);
-  } else {
-    const auto& subTable = subRes->idTable();
-    getExecutionContext()->getSortPerformanceEstimator().throwIfEstimateTooLong(
-        subTable.numRows(), subTable.numColumns(), deadline_, "Sort operation");
-    return computeResultInMemory(subRes);
+  size_t numColumns = subtree_->getResultWidth();
+  // Maximum number of rows that can be sorted in memory.
+  size_t maxNumRowsToBeSortedInMemory =
+      getRuntimeParameter<&RuntimeParameters::sortInMemoryThreshold_>()
+          .getBytes() /
+      (numColumns * sizeof(Id));
+
+  // Always request lazy input to avoid premature materialization.
+  std::shared_ptr<const Result> subRes = subtree_->getResult(true);
+
+  // For fully materialized input, we know the size upfront.
+  if (subRes->isFullyMaterialized()) {
+    const IdTable& table = subRes->idTable();
+    if (table.numRows() <= maxNumRowsToBeSortedInMemory) {
+      return computeResultInMemory(table.clone(),
+                                   subRes->getSharedLocalVocab());
+    } else {
+      std::vector<IdTable> blocks;
+      blocks.push_back(table.clone());
```

Comment:
If you call clone here, there's no benefit of using the external sort. At this point you have two versions of the whole table, one const version and a mutable version, so you could just sort it in memory again. You have to pass it via reference somehow, maybe template the vector or something.